src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,355 @@
+                * greenbidsAnalyticsProperties.getExploratorySamplingSplit();
+        final double throttledSamplingRate = samplingRate
+                * (1.0 - greenbidsAnalyticsProperties.getExploratorySamplingSplit());
+
+        final long hashInt = Integer.parseInt(
+                greenbidsId.substring(greenbidsId.length() - 4), 16);
+        final boolean isPrimarySampled = hashInt < exploratorySamplingRate * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+
+        return isPrimarySampled
+                || hashInt >= (1 - throttledSamplingRate) * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+    }
+
+    private CommonMessage createBidMessage(
+            AuctionContext auctionContext,
+            BidResponse bidResponse,
+            String greenbidsId,
+            String billingId) {
+        final List<Imp> imps = Optional.ofNullable(auctionContext)
+                .map(AuctionContext::getBidRequest)
+                .map(BidRequest::getImp)
+                .orElse(Collections.emptyList());
+
+        if (CollectionUtils.isEmpty(imps)) {
+            throw new IllegalArgumentException("AdUnits list should not be empty");
+        }
+
+        final long auctionElapsed = Optional.ofNullable(auctionContext.getBidRequest())
+                .map(BidRequest::getExt)
+                .map(ExtRequest::getPrebid)
+                .map(ExtRequestPrebid::getAuctiontimestamp)
+                .map(timestamp -> System.currentTimeMillis() - timestamp).orElse(0L);
+
+        final Map<String, Bid> seatsWithBids = Stream.ofNullable(bidResponse.getSeatbid())
+                .flatMap(Collection::stream)
+                .filter(seatBid -> !seatBid.getBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatBid::getSeat,
+                                seatBid -> seatBid.getBid().getFirst(),
+                                (existing, replacement) -> existing));
+
+        final List<SeatNonBid> seatNonBids = auctionContext.getBidRejectionTrackers().entrySet().stream()
+                .map(entry -> toSeatNonBid(entry.getKey(), entry.getValue()))
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .toList();
+
+        final Map<String, NonBid> seatsWithNonBids = Stream.ofNullable(seatNonBids)
+                .flatMap(Collection::stream)
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatNonBid::getSeat,
+                                seatNonBid -> seatNonBid.getNonBid().getFirst(),
+                                (existing, replacement) -> existing));
+
+        final List<AdUnit> adUnitsWithBidResponses = extractAdUnitsWithBidResponses(
+                imps, seatsWithBids, seatsWithNonBids);
+
+        final String auctionId = Optional.ofNullable(auctionContext)
```

Comment:
`Optional.ofNullable(auctionContext).map(AuctionContext::getBidRequest)` can be extracted into the variable and reused three times

even `Optional.ofNullable(auctionContext.getBidRequest())` is fine, because the presense of the bid request is kind of guaranteed